controllers/articles.controllers.js

```diff
@@ -23,16 +24,41 @@ exports.getCommentsByArticleId = (req, res, next) => {
             res.status(200).send({ comments });
         })
         .catch((err) => {
+
             next(err);
         });
 
 }
 
 exports.getAllArticles = (req, res, next) => {
+    const { topic } = req.query
+
+    if (topic) {
+        Promise.all([retrieveAllArticles(topic), retrieveAllTopics(topic)])
+            .then(([articles, topic]) => {
+
+                if (!topic.length && !articles.length) {
+                    res.status(400).send({ msg: 'Invalid topic!' })
+                } else if (!articles.length && topic.length) {
+                    res.status(404).send({ msg: 'No articles found with that topic' })
+                } else {
+                    res.status(200).send({ articles });
+
+                }
+            })
+            .catch((err) => {
+
+                next(err);
+            });
+    } else {
+        retrieveAllArticles().then((articles) => {
+            res.status(200).send({ articles });
+        })
+            .catch((err) => {
+                next(err)
+            })
```

Comment:
A few things here, Ideally we like to keep the controller free of too much logic. Remember the benefit of MVC is that it's partitioned. If we wanted to use a different DB we could, so you don't want too much logic about the db response in here. I completely understand this task is challenging especially with your proposed test suite, so I'm not gonna be too scrupulous.

Secondly, I would avoid sending back errors directly to the client. The convention is to handle errors within error handlers using next. This allows greater scalability and is much handier when it comes to things like monitoring errors or being notified of them etc. 

Try to keep errors handled with custom rejections in those model functions 👍